# Proyek Analisis Data: E-Commerce Public Dataset
- **Nama:** Muh Nur Irfan
- **Email:** Irfanhebat294@gmail.com
- **ID Dicoding:** Muh Nur Irfan

## Menentukan Pertanyaan Bisnis

- Produk apa yang memiliki pembelian terbanyak dan rating terbaik?
- Bagaimana distribusi penjualan produk di berbagai kota?

## Import Semua Packages/Library yang Digunakan

Pada bagian ini saya meimport packages yang ingin digunakan untuk menjalankan semua program selanjutnya juga memberikan as atau alias buat mempermudah dalam pemanggilan packages nya

In [1]:
import pandas as pd   # Untuk manipulasi dan analisis data berbasis tabel
import numpy as np    # Untuk komputasi numerik
!pip install openpyxl


In [2]:
import matplotlib.pyplot as plt   # Untuk membuat visualisasi data dasar
import seaborn as sns
from IPython.display import display
import ipywidgets as widgets

In [3]:
df_customers_dataset=pd.read_csv("customers_dataset.csv")
df_orders_dataset=pd.read_csv("orders_dataset.csv")
df_products_dataset=pd.read_csv("products_dataset.csv")
df_sellers_dataset=pd.read_csv("sellers_dataset.csv")
df_order_items_dataset=pd.read_csv("order_items_dataset.csv")
df_order_payments_dataset=pd.read_csv("order_payments_dataset.csv")
df_order_reviews_dataset=pd.read_csv("order_reviews_dataset.csv")
df_product_category_name_translation=pd.read_csv("product_category_name_translation.csv")
df_geolocation_dataset=pd.read_csv("geolocation_dataset.csv")

## Data Wrangling

### Gathering Data

Gathering Untuk Business Question 1

Pada data pertama saya membuat program untuk menggabungkan 2 data yaitu df_order_items_dataset, df_order_reviews_dataset dengan rujukan order_id lalu diberi nama product_favorit_awal setelah itu saya menggabungkan lagi datanya dengan df_products_dataset dengan rujukan data product_id lalu memberikan nama product_favorit_akhir.sebagai data untuk mendapatkan hasil dari business question pertama saya

In [4]:
product_favorit_awal=pd.merge(df_order_items_dataset, df_order_reviews_dataset, on='order_id')

In [5]:
product_favorit_akhir=pd.merge(product_favorit_awal, df_products_dataset, on='product_id')
print(product_favorit_akhir)
#product_favorit_akhir.to_csv('product_favorit_akhir.csv', index=False)

                                order_id  order_item_id  \
0       00010242fe8c5a6d1ba2dd792cb16214              1   
1       00018f77f2f0320c557190d7a144bdd3              1   
2       000229ec398224ef6ca0657da4fc703e              1   
3       00024acbcdf0a6daa1e931b038114c75              1   
4       00042b26cf59d7ce69dfabb4e55b4fd9              1   
...                                  ...            ...   
112367  fffc94f6ce00a00581880bf54a75a037              1   
112368  fffcd46ef2263f404302a634eb57f7eb              1   
112369  fffce4705a9662cd70adb13d4a31832d              1   
112370  fffe18544ffabc95dfada21779c9644f              1   
112371  fffe41c64501cc87c801fd61db3f6244              1   

                              product_id                         seller_id  \
0       4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1       e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2       c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242a

Gathering Untuk Business Question 2

Pada data kedua ini saya menggunakan data product_favorit_akhir dan df_orders_dataset dengan rujukan data yaitu order_id setelah itu saaya menggabungkan kembali datanya kedalam data baru yang bernama Product_stok_kota_final

In [6]:
Product_stok_kota_awal=pd.merge(product_favorit_akhir, df_orders_dataset, on='order_id')

In [7]:
Product_stok_kota_final=pd.merge(Product_stok_kota_awal, df_customers_dataset, on='customer_id')
print(Product_stok_kota_final)
#Product_stok_kota_final.to_csv('Product_stok_kota_final', index=False)

                                order_id  order_item_id  \
0       00010242fe8c5a6d1ba2dd792cb16214              1   
1       00018f77f2f0320c557190d7a144bdd3              1   
2       000229ec398224ef6ca0657da4fc703e              1   
3       00024acbcdf0a6daa1e931b038114c75              1   
4       00042b26cf59d7ce69dfabb4e55b4fd9              1   
...                                  ...            ...   
112367  fffc94f6ce00a00581880bf54a75a037              1   
112368  fffcd46ef2263f404302a634eb57f7eb              1   
112369  fffce4705a9662cd70adb13d4a31832d              1   
112370  fffe18544ffabc95dfada21779c9644f              1   
112371  fffe41c64501cc87c801fd61db3f6244              1   

                              product_id                         seller_id  \
0       4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1       e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2       c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242a

**Insight business question 1:**
- Dengan mengaitkan 3 data yang terdapat pada order item data set,order review data,dan product dataset dalam product favorit akhir maka saya mempunyai data yang saya butuhkan untuk melihat produk dengan pembelian terbanyak dan terfavorit dari data tersebut.(product_favorit_akhir)

**Insight business question 2:**
- Dengan mengaitkan data pada product akhir dengan dua data setelah nya yaitu orders dataset dan costumers dataset menjadi data product stok kota final maka saya mempunya data yang terkait dan dapat mengetahui barang paling banyak keluar disetiap kotanya(Product_stok_kota_final)

### Assessing Data

Pada bagian ini saya melakukan Assesing data terutama pada bagian missing value untuk melihat berapa missing value yang terdapat pada data saya tertutama pada kolom yang ingin saya gunakan

In [8]:
product_favorit_akhir.isna().sum()

order_id                          0
order_item_id                     0
product_id                        0
seller_id                         0
shipping_limit_date               0
price                             0
freight_value                     0
review_id                         0
review_score                      0
review_comment_title          98938
review_comment_message        64730
review_creation_date              0
review_answer_timestamp           0
product_category_name          1598
product_name_lenght            1598
product_description_lenght     1598
product_photos_qty             1598
product_weight_g                 18
product_length_cm                18
product_height_cm                18
product_width_cm                 18
dtype: int64

In [9]:
Product_stok_kota_final.isna().sum()

order_id                             0
order_item_id                        0
product_id                           0
seller_id                            0
shipping_limit_date                  0
price                                0
freight_value                        0
review_id                            0
review_score                         0
review_comment_title             98938
review_comment_message           64730
review_creation_date                 0
review_answer_timestamp              0
product_category_name             1598
product_name_lenght               1598
product_description_lenght        1598
product_photos_qty                1598
product_weight_g                    18
product_length_cm                   18
product_height_cm                   18
product_width_cm                    18
customer_id                          0
order_status                         0
order_purchase_timestamp             0
order_approved_at                   15
order_delivered_carrier_date      1184
order_delivered_customer_date     2360
order_estimated_delivery_date        0
customer_unique_id                   0
customer_zip_code_prefix             0
customer_city                        0
customer_state                       0
dtype: int64

Pada bagian ini saya melakukan IQR yang biasa digunakan untuk melihat data outlier dan membuatkan ambang batas untuk data tertinggi dan data ter rendah buat di analisis agar tidak mempengaruhi hasil dari analisis data selanjutnya

In [10]:
# Hitung Q1 dan Q3 secara eksplisit
Q1 = np.percentile(product_favorit_akhir['review_score'], 25)
Q3 = np.percentile(product_favorit_akhir['review_score'], 75)

# Hitung IQR
IQR = Q3 - Q1

# Hitung lower bound dan upper bound
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

print(f"Q1: {Q1}, Q3: {Q3}")
print(f"IQR: {IQR}")
print(f"Lower bound: {lower_bound}, Upper bound: {upper_bound}")


Q1: 4.0, Q3: 5.0
IQR: 1.0
Lower bound: 2.5, Upper bound: 6.5


**Insight business question 1:**
- Pada data pertama kolom yang ingin saya gunakan yaitu review score dan product category dimana insight yang saya dapatkan di tabel product_category_name terdapat missing value (1598)

**Insight business question 2:**
- Pada data kedua kolom yang saya ingin saya gunakan adalah data pada product_category_name dan customer_city dan customer_state dimana masalah pada missing value nya juga sama yaitu terdapat pada data product_category_name

### Cleaning Data

Pada bagian ini saya melakukan cleaning data terutama pada kolom utama yang ingin saya gunakan yaitu Product_category_name pada setiap data yang saya gunakan agar dapat mendapatkan hasil analisis yang digunakan

In [11]:
product_favorit_akhir.dropna(subset=['product_category_name'], inplace=True)
Product_stok_kota_final.dropna(subset=['product_category_name'], inplace=True)

In [12]:
Product_stok_kota_final.isna().sum()

order_id                             0
order_item_id                        0
product_id                           0
seller_id                            0
shipping_limit_date                  0
price                                0
freight_value                        0
review_id                            0
review_score                         0
review_comment_title             97418
review_comment_message           63839
review_creation_date                 0
review_answer_timestamp              0
product_category_name                0
product_name_lenght                  0
product_description_lenght           0
product_photos_qty                   0
product_weight_g                     1
product_length_cm                    1
product_height_cm                    1
product_width_cm                     1
customer_id                          0
order_status                         0
order_purchase_timestamp             0
order_approved_at                   14
order_delivered_carrier_date      1147
order_delivered_customer_date     2295
order_estimated_delivery_date        0
customer_unique_id                   0
customer_zip_code_prefix             0
customer_city                        0
customer_state                       0
dtype: int64

**Insight:**
- Pada data satu dan data kedua saya menggunakan dropna karena menurut saya data yang dihapus tidak akan mempengaruhi hasil dari insight terakhir karena hanya memiliki sedikit missing value yang tidak akan mempengaruhi banyak nya sebaran data yang ada
- Pada data dua juga saya menggunakan dropna untuk menghapus sebaran missing value pada kolom tersebut

## Exploratory Data Analysis (EDA)

### Explore ...

**Eploratory Data Analysis Business question 1**

Program ini dimulai dengan memverifikasi apakah data yang digunakan memiliki kolom yang diperlukan, yaitu `product_category_name`, `review_score`, dan `order_item_id`. Setelah itu, data dikelompokkan berdasarkan kategori produk (`product_category_name`), dengan menghitung jumlah pembelian (`purchase_count`) dan rata-rata rating (`average_rating`) per kategori. Kemudian, data diurutkan dua kali: pertama berdasarkan jumlah pembelian terbanyak, dan kedua berdasarkan rating tertinggi. Hasil kedua urutan ini digabungkan untuk mendapatkan produk yang muncul di kedua kategori tersebut. Selanjutnya, program menampilkan lima produk teratas yang memiliki kombinasi pembelian terbanyak dan rating tertinggi, serta memberikan statistik deskriptif untuk kolom `purchase_count` dan `average_rating`. Terakhir, data ditampilkan dengan format yang lebih mudah dibaca, di mana jumlah pembelian dipisahkan dengan ribuan dan rating ditampilkan dengan dua angka desimal.

In [26]:
# Memeriksa apakah kolom yang diperlukan ada
required_columns = ["product_category_name", "review_score", "order_item_id"]
if not all(col in product_favorit_akhir.columns for col in required_columns):
    raise ValueError(f"Data harus memiliki kolom: {required_columns}")

# Mengelompokkan data berdasarkan kategori produk dan menghitung pembelian unik berdasarkan order_item_id
grouped_data = product_favorit_akhir.groupby("product_category_name").agg(
    purchase_count=("order_item_id", "sum"),  # Menghitung jumlah pembelian
    average_rating=("review_score", "mean")  # Menghitung rata-rata rating
).reset_index()

# Menyaring produk dengan pembelian terbanyak dan rating tertinggi
# Mengurutkan berdasarkan pembelian terbanyak
top_purchased_products = grouped_data.sort_values(by="purchase_count", ascending=False)

# Mengurutkan berdasarkan rating tertinggi
top_rated_products = grouped_data.sort_values(by="average_rating", ascending=False)

# Menampilkan produk yang ada di kedua kategori: pembelian terbanyak dan rating tertinggi
# Produk yang ada di dua urutan teratas (terbaik)
top_combined_products = pd.merge(top_purchased_products[['product_category_name', 'purchase_count']],
                                  top_rated_products[['product_category_name', 'average_rating']],
                                  on="product_category_name")

# Menampilkan 5 produk dengan pembelian terbanyak dan rating tertinggi
top_combined_products[['product_category_name', 'purchase_count', 'average_rating']].head()

# Menampilkan statistik deskriptif untuk kolom numerik
top_combined_products[['purchase_count', 'average_rating']].describe()

# Menampilkan data dengan format yang lebih baik
top_combined_products[['product_category_name', 'purchase_count', 'average_rating']].style.format({
    'purchase_count': '{:,}',  # Menampilkan jumlah pembelian dengan pemisah ribuan
    'average_rating': '{:.2f}'  # Menampilkan rating dengan dua angka desimal
})


- Program ini bertujuan untuk menganalisis produk berdasarkan dua metrik utama: jumlah pembelian dan rata-rata rating. Pertama, program memeriksa apakah data memiliki kolom yang diperlukan, yaitu `product_category_name`, `review_score`, dan `order_item_id`. Selanjutnya, data dikelompokkan berdasarkan kategori produk dan dihitung jumlah pembelian serta rata-rata rating per kategori. Setelah itu, data diurutkan berdasarkan jumlah pembelian terbanyak dan rating tertinggi. Program kemudian menggabungkan dua urutan ini untuk menampilkan produk yang muncul di kedua kategori tersebut. Selanjutnya, produk yang memiliki jumlah pembelian lebih dari 5000 dan rating di atas 4 disaring, dan hanya 5 produk teratas yang memenuhi kriteria tersebut yang ditampilkan bersama dengan jumlah pembelian dan rating rata-rata masing-masing.

In [35]:

# Memeriksa apakah kolom yang diperlukan ada
required_columns = ["product_category_name", "review_score", "order_item_id"]
if not all(col in product_favorit_akhir.columns for col in required_columns):
    raise ValueError(f"Data harus memiliki kolom: {required_columns}")

# Mengelompokkan data berdasarkan kategori produk dan menghitung pembelian unik berdasarkan order_item_id
grouped_data = product_favorit_akhir.groupby("product_category_name").agg(
    purchase_count=("order_item_id", "sum"),  # Menghitung jumlah pembelian
    average_rating=("review_score", "mean")  # Menghitung rata-rata rating
).reset_index()

# Menyaring produk dengan pembelian terbanyak dan rating tertinggi
# Mengurutkan berdasarkan pembelian terbanyak
top_purchased_products = grouped_data.sort_values(by="purchase_count", ascending=False)

# Mengurutkan berdasarkan rating tertinggi
top_rated_products = grouped_data.sort_values(by="average_rating", ascending=False)

# Menampilkan produk yang ada di kedua kategori: pembelian terbanyak dan rating tertinggi
top_combined_products = pd.merge(top_purchased_products[['product_category_name', 'purchase_count']],
                                  top_rated_products[['product_category_name', 'average_rating']],
                                  on="product_category_name")

# Menyaring produk dengan purchase_count > 5000 dan average_rating > 4
filtered_products = top_combined_products[(top_combined_products['purchase_count'] > 5000) &
                                           (top_combined_products['average_rating'] > 4)]

# Menampilkan 5 produk dengan purchase_count tertinggi yang memenuhi kriteria
top_5_purchased_products = filtered_products.head(5)

# Menampilkan hasil
top_5_purchased_products[['product_category_name', 'purchase_count', 'average_rating']]


product_category_name  purchase_count  average_rating
2           beleza_saude           11027        4.142768
3          esporte_lazer            9928        4.107986
5  utilidades_domesticas            9027        4.055019
6     relogios_presentes            6548        4.019160
7     ferramentas_jardim            5847        4.042735

-EDA BUSINESS QUESTION 2

- Program ini bertujuan untuk menganalisis data pembelian berdasarkan kota. Pertama, program memeriksa apakah kolom yang diperlukan, yaitu "customer_city" dan "order_item_id", ada dalam dataset. Kemudian, data dikelompokkan berdasarkan kota, dan jumlah total pembelian dihitung untuk setiap kota dengan menggunakan `groupby` dan fungsi agregasi `count` pada kolom `order_item_id`. Setelah itu, data diurutkan berdasarkan jumlah pembelian terbanyak, dan 10 kota dengan pembelian tertinggi dipilih. Hasilnya, program menampilkan daftar 10 kota dengan jumlah pembelian terbanyak, di mana jumlah pembelian untuk setiap kota ditampilkan dengan format pemisah ribuan untuk mempermudah pembacaan data.

In [42]:

# Memeriksa apakah kolom yang diperlukan ada
required_columns = ["customer_city", "order_item_id"]
if not all(col in Product_stok_kota_final.columns for col in required_columns):
    raise ValueError(f"Data harus memiliki kolom: {required_columns}")

# Mengelompokkan data berdasarkan kota dan menghitung jumlah total pembelian (order_item_id)
grouped_data = Product_stok_kota_final.groupby("customer_city").agg(
    total_purchases=("order_item_id", "count")  # Menghitung jumlah pembelian per kota
).reset_index()

# Mengurutkan data berdasarkan jumlah pembelian terbanyak
sorted_data = grouped_data.sort_values(by="total_purchases", ascending=False)

# Menampilkan 10 kota dengan jumlah pembelian terbanyak
top_10_cities = sorted_data.head(10)

# Menampilkan hasil
top_10_cities[['customer_city', 'total_purchases']].style.format({
    'total_purchases': '{:,}'  # Menampilkan jumlah pembelian dengan pemisah ribuan
})


- Program ini menganalisis data pembelian berdasarkan kota dan kategori produk. Langkah pertama adalah memeriksa apakah kolom yang diperlukan, yaitu "customer_city", "order_item_id", dan "product_category_name", ada dalam dataset. Selanjutnya, data dikelompokkan berdasarkan kota untuk menghitung jumlah pembelian total (dihitung dengan menghitung jumlah `order_item_id`), yang kemudian diurutkan berdasarkan jumlah pembelian terbanyak. Dari data yang sudah terurut, program memilih 10 kota dengan jumlah pembelian tertinggi. Data ini kemudian disaring untuk hanya mencakup kota-kota teratas tersebut. Setelah itu, program mengelompokkan data berdasarkan kota dan kategori produk untuk menghitung jumlah pembelian per kategori produk di setiap kota. Kemudian, program menyaring produk dengan jumlah pembelian terbanyak di setiap kota menggunakan fungsi `nlargest`. Hasil akhirnya adalah menampilkan kota-kota teratas beserta kategori produk yang memiliki jumlah pembelian tertinggi di masing-masing kota, dengan format pemisah ribuan pada jumlah pembelian agar lebih mudah dibaca.

In [ ]:
import pandas as pd

# Memeriksa apakah kolom yang diperlukan ada
required_columns = ["customer_city", "order_item_id", "product_category_name"]
if not all(col in Product_stok_kota_final.columns for col in required_columns):
    raise ValueError(f"Data harus memiliki kolom: {required_columns}")

# Mengelompokkan data berdasarkan kota dan menghitung jumlah total pembelian (order_item_id)
grouped_data = Product_stok_kota_final.groupby("customer_city").agg(
    total_purchases=("order_item_id", "count")  # Menghitung jumlah pembelian per kota
).reset_index()

# Mengurutkan data berdasarkan jumlah pembelian terbanyak
sorted_data = grouped_data.sort_values(by="total_purchases", ascending=False)

# Menampilkan 10 kota dengan jumlah pembelian terbanyak
top_10_cities = sorted_data.head(10)

# Menyaring data hanya untuk 10 kota teratas
top_10_cities_data = Product_stok_kota_final[Product_stok_kota_final['customer_city'].isin(top_10_cities['customer_city'])]

# Mengelompokkan data berdasarkan kota dan kategori produk, dan menghitung jumlah pembelian per kategori produk
grouped_products = top_10_cities_data.groupby(["customer_city", "product_category_name"]).agg(
    total_product_sales=("order_item_id", "count")  # Menghitung jumlah pembelian per kategori produk di setiap kota
).reset_index()

# Menyaring produk dengan pembelian terbanyak di setiap kota menggunakan nlargest
top_products_per_city = grouped_products.groupby("customer_city").apply(
    lambda x: x.nlargest(1, 'total_product_sales')
).reset_index(drop=True)

# Menampilkan hasil
top_products_per_city[['customer_city', 'product_category_name', 'total_product_sales']].style.format({
    'total_product_sales': '{:,}'  # Menampilkan jumlah pembelian dengan pemisah ribuan
})


**Insight business question 1:**
- Dari hasil analisis yang dilakukan, kita dapat menarik beberapa wawasan penting. Pertama, produk dengan jumlah pembelian tinggi dan rating yang sangat baik (di atas 5000 pembelian dan rating lebih dari 4) cenderung menunjukkan bahwa produk tersebut populer dan disukai oleh banyak pelanggan. Hal ini mencerminkan bahwa kualitas produk tersebut diakui oleh konsumen, dan tingkat kepuasan yang tinggi berkontribusi terhadap volume penjualan yang besar. Produk-produk yang memenuhi kriteria ini dapat dianggap sebagai produk unggulan yang memiliki performa terbaik di pasar, baik dari sisi volume penjualan maupun reputasi pelanggan. Mengetahui produk teratas ini bisa membantu perusahaan dalam merencanakan strategi pemasaran lebih lanjut, memperkuat penawaran yang ada, dan memfokuskan upaya pengembangan produk pada kategori yang paling diminati oleh konsumen.

**Insight business question 2:**
- Berdasarkan analisis data pembelian ini, kita dapat melihat kota-kota dengan jumlah pembelian tertinggi dan kategori produk yang paling diminati di masing-masing kota tersebut. Hal ini memberikan wawasan penting bagi perusahaan atau penjual untuk lebih fokus pada kota-kota yang memiliki tingkat pembelian yang lebih tinggi, yang mungkin menunjukkan pasar yang lebih potensial. Selain itu, informasi tentang kategori produk dengan pembelian terbanyak di setiap kota memungkinkan penjual untuk mengidentifikasi produk yang lebih diminati oleh konsumen di wilayah tertentu. Dengan memahami preferensi konsumen berdasarkan lokasi dan kategori produk, perusahaan dapat merancang strategi pemasaran dan stok yang lebih tepat sasaran, serta meningkatkan pengalaman pelanggan dengan menawarkan produk yang lebih relevan.

## Visualization & Explanatory Analysis

### Pertanyaan 1:

Program ini dimulai dengan memverifikasi apakah data yang digunakan memiliki kolom yang diperlukan, yaitu product_category_name, review_score, dan order_item_id. Setelah memastikan keberadaan kolom tersebut, data kemudian dikelompokkan berdasarkan kategori produk (product_category_name) untuk menghitung total pembelian (purchase_count) dan rata-rata rating (average_rating). Data yang telah dikelompokkan ini kemudian disortir untuk mendapatkan produk dengan pembelian terbanyak dan rating tertinggi. Selanjutnya, program menggabungkan dua hasil sorting tersebut untuk mendapatkan produk yang memiliki pembelian lebih dari 5000 dan rating lebih dari 4. Program akhirnya menampilkan lima produk teratas yang memenuhi kriteria tersebut dalam bentuk visualisasi bar chart interaktif menggunakan Plotly, yang memungkinkan pengguna untuk melihat total pembelian dan rating produk dengan lebih jelas.

In [30]:
import pandas as pd
import plotly.express as px

# Memeriksa apakah kolom yang diperlukan ada
required_columns = ["product_category_name", "review_score", "order_item_id"]
if not all(col in product_favorit_akhir.columns for col in required_columns):
    raise ValueError(f"Data harus memiliki kolom: {required_columns}")

# Mengelompokkan data berdasarkan kategori produk dan menghitung pembelian unik berdasarkan order_item_id
grouped_data = product_favorit_akhir.groupby("product_category_name").agg(
    purchase_count=("order_item_id", "sum"),  # Menghitung jumlah pembelian
    average_rating=("review_score", "mean")  # Menghitung rata-rata rating
).reset_index()

# Menyaring produk dengan pembelian terbanyak dan rating tertinggi
top_purchased_products = grouped_data.sort_values(by="purchase_count", ascending=False)
top_rated_products = grouped_data.sort_values(by="average_rating", ascending=False)

# Menampilkan produk yang ada di kedua kategori: pembelian terbanyak dan rating tertinggi
top_combined_products = pd.merge(top_purchased_products[['product_category_name', 'purchase_count']],
                                  top_rated_products[['product_category_name', 'average_rating']],
                                  on="product_category_name")

# Menyaring produk dengan purchase_count > 5000 dan average_rating > 4
filtered_products = top_combined_products[(top_combined_products['purchase_count'] > 5000) &
                                           (top_combined_products['average_rating'] > 4)]

# Menampilkan 5 produk dengan purchase_count tertinggi yang memenuhi kriteria
top_5_purchased_products = filtered_products.head(5)

# Membuat visualisasi interaktif menggunakan Plotly
fig = px.bar(top_5_purchased_products,
             x='purchase_count',
             y='product_category_name',
             text='purchase_count',  # Menambahkan nilai pada bar
             hover_data={'purchase_count': True, 'average_rating': True},  # Menambahkan data saat hover
             labels={'purchase_count': 'Purchase Count', 'product_category_name': 'Product Category'},
             title="Top 5 Products with Highest Purchase Count and Rating Above 4")

# Menampilkan grafik dengan interaktif
fig.update_traces(texttemplate='%{text}', textposition='inside', hoverinfo="x+y+text")
fig.show()


### Pertanyaan 2:

Program ini menganalisis data penjualan berdasarkan kota dan kategori produk dari dataset Product_stok_kota_final. Proses dimulai dengan memeriksa apakah kolom yang diperlukan ada dalam data. Kemudian, data dikelompokkan berdasarkan kota untuk menghitung jumlah pembelian per kota. Setelah itu, program mengurutkan kota berdasarkan total pembelian terbanyak dan memilih 10 kota teratas. Selanjutnya, data disaring untuk kota-kota tersebut, dan pembelian per kategori produk dihitung. Produk dengan pembelian terbanyak di setiap kota dipilih, dan akhirnya, visualisasi interaktif dibuat menggunakan Plotly, menampilkan total penjualan produk di setiap kota dengan informasi tambahan saat hover. Hasilnya adalah grafik batang yang memperlihatkan produk terlaris per kota.

In [50]:
import pandas as pd
import plotly.express as px

# Memeriksa apakah kolom yang diperlukan ada
required_columns = ["customer_city", "order_item_id", "product_category_name"]
if not all(col in Product_stok_kota_final.columns for col in required_columns):
    raise ValueError(f"Data harus memiliki kolom: {required_columns}")

# Mengelompokkan data berdasarkan kota dan menghitung jumlah total pembelian (order_item_id)
grouped_data = Product_stok_kota_final.groupby("customer_city").agg(
    total_purchases=("order_item_id", "count")  # Menghitung jumlah pembelian per kota
).reset_index()

# Mengurutkan data berdasarkan jumlah pembelian terbanyak
sorted_data = grouped_data.sort_values(by="total_purchases", ascending=False)

# Menampilkan 10 kota dengan jumlah pembelian terbanyak
top_10_cities = sorted_data.head(10)

# Menyaring data hanya untuk 10 kota teratas
top_10_cities_data = Product_stok_kota_final[Product_stok_kota_final['customer_city'].isin(top_10_cities['customer_city'])]

# Mengelompokkan data berdasarkan kota dan kategori produk, dan menghitung jumlah pembelian per kategori produk
grouped_products = top_10_cities_data.groupby(["customer_city", "product_category_name"]).agg(
    total_product_sales=("order_item_id", "count")  # Menghitung jumlah pembelian per kategori produk di setiap kota
).reset_index()

# Menyaring produk dengan pembelian terbanyak di setiap kota menggunakan nlargest
top_products_per_city = grouped_products.groupby("customer_city").apply(
    lambda x: x.nlargest(1, 'total_product_sales')
).reset_index(drop=True)

# Membuat diagram interaktif dengan Plotly
fig = px.bar(
    top_products_per_city,
    x='total_product_sales',
    y='customer_city',
    color='product_category_name',
    title='Top Products by City',
    labels={'total_product_sales': 'Total Product Sales', 'customer_city': 'Customer City', 'product_category_name': 'Product Category'},
    text='total_product_sales',  # Menambahkan nilai total pembelian sebagai teks di atas batang
    hover_data={'customer_city': True, 'product_category_name': True, 'total_product_sales': True}  # Menampilkan data saat hover
)

# Menampilkan plot interaktif
fig.update_traces(textposition='outside', texttemplate='%{text}', hoverinfo='x+y+text+name')

fig.update_layout(
    xaxis_title="Total Product Sales",
    yaxis_title="Customer City",
    template="plotly_dark"
)

fig.show()


<ipython-input-50-0484d76d31dd>:29: DeprecationWarning:

DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.



**Insight business question 1:**
- Dari analisis ini, kita dapat melihat bahwa produk dengan jumlah pembelian tinggi tidak selalu memiliki rating tertinggi, namun produk dengan pembelian lebih dari 5000 dan rating lebih dari 4 cenderung menunjukkan kualitas yang baik dan diterima dengan baik oleh konsumen. Hal ini memberi indikasi bahwa perusahaan harus lebih fokus pada produk-produk yang tidak hanya populer dalam hal jumlah pembelian, tetapi juga mendapat ulasan positif dari pelanggan. Dengan memahami kombinasi ini, perusahaan bisa memprioritaskan produk yang memenuhi kedua kriteria tersebut untuk strategi stok dan pemasaran yang lebih efektif, serta memperkuat kepuasan pelanggan.

**Insight business question 2:**
- Insight dari analisis ini menunjukkan bahwa terdapat variasi dalam preferensi produk di berbagai kota, dengan beberapa kota memiliki produk tertentu yang lebih banyak dibeli dibandingkan kota lainnya. Melalui visualisasi ini, kita dapat mengidentifikasi produk yang memiliki permintaan tinggi di masing-masing kota, yang dapat menjadi dasar untuk strategi pengelolaan stok yang lebih efisien. Misalnya, jika suatu produk memiliki jumlah penjualan yang tinggi di banyak kota, maka perusahaan dapat mempertimbangkan untuk menambah stok produk tersebut di kota-kota yang menunjukkan tren permintaan tinggi. Selain itu, informasi ini membantu dalam merencanakan kampanye pemasaran yang lebih terfokus pada produk-produk yang populer di wilayah tertentu..

## Analisis Lanjutan (Opsional)

## Conclusion

**Conclution pertanyaan 1**
- Dari analisis yang pertama, kita fokus pada jumlah pembelian setiap produk dan menetapkan aturan untuk menyaring produk yang memiliki jumlah pembelian minimal 5000 unit dan rating rata-rata di atas 4. Hasilnya, lima produk yang memenuhi kriteria tersebut adalah **beleza_saude**, **esporte_lazer**, **utilidades_domesticas**, **relogios_presentes**, dan **ferramentas_jardim**. Produk-produk ini tidak hanya memiliki volume penjualan yang tinggi, tetapi juga mendapat rating yang sangat baik dari pelanggan, menunjukkan bahwa mereka sangat dihargai dan diterima baik oleh konsumen. Oleh karena itu, produk-produk ini dapat dianggap sebagai prioritas dalam strategi pemasaran dan pengelolaan stok. Dengan memfokuskan upaya pada produk-produk ini, perusahaan dapat meningkatkan kepuasan pelanggan dan memastikan ketersediaan barang yang dibutuhkan oleh pasar.

  Secara keseluruhan, analisis ini memberikan wawasan yang berharga bagi perusahaan untuk merancang strategi operasional dan pemasaran yang lebih terfokus dan efisien, dengan memperhatikan baik aspek popularitas produk maupun kualitas yang tercermin dalam rating pelanggan.

**Conclution pertanyaan 2**
- Berdasarkan analisis yang dilakukan, ditemukan bahwa preferensi produk bervariasi antara kota-kota, dengan beberapa produk lebih populer dan memiliki penjualan yang lebih tinggi di kota tertentu. Hal ini menunjukkan pentingnya memahami pola permintaan lokal untuk mengoptimalkan pengelolaan stok dan strategi pemasaran. Perusahaan dapat menggunakan informasi ini untuk menambah stok produk yang memiliki permintaan tinggi di kota-kota tertentu, sehingga dapat mengurangi risiko kehabisan stok dan meningkatkan kepuasan pelanggan. Selain itu, wawasan ini juga memungkinkan perusahaan merancang kampanye pemasaran yang lebih terfokus, dengan menyoroti produk-produk yang paling populer di masing-masing wilayah, sehingga dapat menarik lebih banyak konsumen.